File: benchmarks/ultralytics_yolo.exs
[eps, model_path] = System.argv()

eps = case eps do
  "directml" -> [:directml]
  "coreml" -> [:coreml]
  "cuda" -> [:cuda]
  "tensorrt" -> [:tensorrt]
  _ -> [:cpu]
end


mat = Evision.imread("benchmarks/images/traffic.jpg")

classes_path = "models/coco_classes.json"

model = YOLO.load(model_path: model_path, classes_path: classes_path, eps: eps)

dbg(Nx.default_backend())

### warmup
yolo_pipeline = fn mat ->
  {input, scaling_config} = YOLO.Models.Ultralytics.preprocess(model, mat, frame_scaler: YOLO.FrameScalers.EvisionScaler)
  model_input = Nx.backend_copy(input, EXLA.Backend)
  output = YOLO.Models.run(model, input)
  detections = YOLO.Models.Ultralytics.postprocess(model, output, scaling_config, prob_threshold: 0.25, iou_threshold: 0.45)
  {model_input, scaling_config, output, detections}
end

for _ <- 1..50 do
  yolo_pipeline.(mat)
end

{model_input, scaling_config, model_output, _detections} = yolo_pipeline.(mat)

Benchee.run(%{
    "preprocess" => fn ->
      YOLO.Models.Ultralytics.preprocess(model, mat, frame_scaler: YOLO.FrameScalers.EvisionScaler)
    end,
    "run" => {
      fn input -> YOLO.Models.run(model, input) end,
      before_each: fn _ ->
        Nx.backend_copy(model_input, EXLA.Backend)
      end
    },
    "postprocess" => fn ->
      YOLO.Models.Ultralytics.postprocess(model, model_output, scaling_config, prob_threshold: 0.25, iou_threshold: 0.45)
    end
  },
  time: 10,
  memory_time: 2
)

File: benchmarks/ultralytics_yolo.py
from ultralytics import YOLO
import sys
import time

model_path = sys.argv[1]
# default to cpu, can be "mps" or "cuda"
device = sys.argv[2] if len(sys.argv) > 2 else "cpu"

model = YOLO(model_path)

total = 0
count = 0
min_time = 10
max_time = -1

# warmup
for i in range(1, 10):
  model(["benchmarks/images/traffic.jpg"], device=device)
  # Mac
  # model(["benchmarks/images/traffic.jpg"], device='mps')
  
  # CUDA
  # model(["benchmarks/images/traffic.jpg"], device='cuda')

for i in range(0, 10):
  start = time.time()
  model(["benchmarks/images/traffic.jpg"])
  stop = time.time()
  diff = stop - start
  count += 1
  total += diff
  if min_time > diff:
    min_time = diff
  if max_time < diff:
    max_time = diff

print("count: ", count)
print("avg: ", total/count)
print("min: ", min_time)
print("max: ", max_time)
